tree/src/tree.rs

```diff
@@ -0,0 +1,233 @@
+use std::fmt;
+use std::fmt::{Debug, Display};
+
+use arith::{Field, FieldSerde};
+use ark_std::{end_timer, log2, start_timer};
+use rayon::iter::{IndexedParallelIterator, IntoParallelRefMutIterator, ParallelIterator};
+
+use crate::{Leaf, Node, Path};
+
+/// Represents a Merkle tree structure.
+#[derive(Clone, Debug, PartialEq)]
+pub struct Tree<F: Field + FieldSerde> {
+    pub nodes: Vec<Node>,
+    pub leaves: Vec<Leaf<F>>,
+}
+
+impl<F: Field + FieldSerde> Display for Tree<F> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        writeln!(f, "nodes:")?;
+        for (i, e) in self.nodes.iter().enumerate() {
+            writeln!(f, "{}: {}", i, e)?;
+        }
+        writeln!(f, "leaves:")?;
+        for (i, e) in self.leaves.iter().enumerate() {
+            writeln!(f, "{}: {}", i, e)?;
+        }
+        Ok(())
+    }
+}
+
+impl<F: Field + FieldSerde> Tree<F> {
+    /// Creates an empty tree with default leaves.
+    #[inline]
+    pub fn init(tree_height: usize) -> Self {
+        let leaves = vec![Leaf::default(); 1 << (tree_height - 1)];
+        Self::new_with_leaves(leaves)
+    }
+
+    /// Builds a tree with the given leaves.
+    #[inline]
+    pub fn new_with_field_elements(leave_elems: &[F]) -> Self {
+        let leaves = leave_elems
+            .iter()
+            .map(|&leaf| Leaf { data: leaf })
+            .collect::<Vec<Leaf<F>>>();
+        Self::new_with_leaves(leaves)
+    }
+
+    /// Builds a tree with the given leaves.
+    #[inline]
+    pub fn new_with_leaves(leaves: Vec<Leaf<F>>) -> Self {
+        let tree_height = log2(leaves.len() + 1);
```

Comment:
i felt like this can be wrong - consider 7 leaves, then you need 8 leaves capacity, while the height from 7 leaves is different from the one from 8 leaves

Reply:
maybe they are assuming that the leaves are always being po2, s.t., the height is correct in this case